lgc/patch/PatchInOutImportExport.cpp

```diff
@@ -144,6 +155,45 @@ bool PatchInOutImportExport::runOnModule(Module &module) {
   return true;
 }
 
+// =====================================================================================================================
+// Mark the 'done' flag to the very last position export instruction.
+//
+// @param [in/out] func : LLVM function to be run on
+// @param [in] pdt : The PostDominatorTree of the \p func
+void PatchInOutImportExport::markExportDone(Function *func, PostDominatorTree &pdt) {
+  std::vector<CallInst *> expInstrs;
+
+  Function *expDecl = m_module->getFunction("llvm.amdgcn.exp.f32");
+  if (!expDecl)
+    return;
+
+  // get the export call instructions
+  for (auto user : expDecl->users()) {
+    if (CallInst *callInst = dyn_cast<CallInst>(user)) {
+      if (callInst->getFunction() == func) {
+        if (ConstantInt *target = dyn_cast<ConstantInt>(callInst->getOperand(0))) {
+          uint64_t targetValue = target->getZExtValue();
+          if (targetValue >= EXP_TARGET_POS_0 && targetValue <= EXP_TARGET_POS_3)
+            expInstrs.push_back(callInst);
+        }
+      }
+    }
+  }
+
+  if (expInstrs.empty())
+    return;
+
+  CallInst *lastExport = expInstrs[0];
+
+  // Assume the last export post-dominates all other position exports.
+  // I am not sure whether there may be some case breaking this.
```

Comment:
Can anyone else comment on this? In theory I guess there could be no export that post-dominates all other exports, so in that case would you have to add some kind of dummy "export done" at the end of the shader?

Reply:
@jayfoad If you said is true, we have already done seriously wrong with existing logic. @nhaehnle @trenouf  Do you know any graphics program can break the assumption?

Reply:
BTW, a graphics program may have no position export, which is already handled by visitReturnInst(), I am happy to hear any other case that "a program with position exports, but no position export post-dominates others".

Reply:
>  In theory I guess there could be no export that post-dominates all other exports, so in that case would you have to add some kind of dummy "export done" at the end of the shader?

We add dummy export because the shader doesn't export any position data (not because it is partially exported). We construct a return block in lower phase and all exit points must go to this block. Meanwhile, we create a "alloca" pointer to store position data as proxy variable. All "gl_Position = " statements are translated to LLVM IR store instruction to this proxy variable. In the return block, we load position data from this proxy and export it via export intrinsic. If there is at least one "gl_Position = " statement, we will have this proxy variable and have corresponding export intrinsic for position. If there is no "gl_Position" statement, we have to add dummy position export in patch phase.

Reply:
I believe @amdrexu is right here. Please adjust the comment, and add a check that one call dominates the other, i.e.
```
  if (pdt.dominates(expInstrs[i], lastExport)) {
    lastExport = expInstrs[i];
  } else {
    assert(pdt.dominates(lastExport, expInstrs[i]));
  }
```